### ISSUE TITLE
Encrypted files after moving from personal folder to group folder

### ISSUE THREAD

**Issue by Bust011r** [nextcloud/groupfolders#1896]

**Comment by CarlSchwan** [comment]
Very strange issue, it should be impossible to encrypt groupfolders. I will try to reproduce. Do you have a backup?

**Comment by Bust011r** [comment]
unfortunately i didn't have a backup of those files, can filekey be derived somehow?

I hope you can reproduce the problem but I couldn't, so I don't know exactly what caused this bug. I think the move was successful and then they are encrypted after a few days because in the cache I have the thumbnails of the files on the contrary, if I move them now, the preview of the encrypted text remains in the thumbnail.

**Comment by kleener** [comment]
I have the same problem.
Moving the file back helped me.
At the moment I help myself by downloading the files and then uploading them to the group folder.

**Comment by kleener** [comment]
It seems that encryption was added in the last release.
Does anyone have experience with this?
Do you have to activate it first?

https://github.com/nextcloud/groupfolders/releases/tag/v12.0.2

**Comment by PVince81** [comment]
so we need two things:
- [ ] fix the bug where moving into a non-encrypted group folder doesn't decrypt. Probably this is a "moveFromStorage" that should be a "copyFromStorage" call somewhere under those conditions (we usually use "moveFromStorage" even cross-storages to save performance)
- [ ] extend the repair command to fix issues caused by the bug

**Comment by PVince81** [comment]
I've stepped with a debugger and it lead to this line: https://github.com/nextcloud/server/blob/stable24/lib/private/Files/Storage/Local.php#L587

This is the shortcut the code takes when moving from a `Local` storage to another `Local` storage, the group folder is stored on that. This also tells me that if a primary object store was used, it is likely that the problem doesn't happen there.

We need to expand that code somehow to not take the shortcut whenever we detect that the encryption state of source and target storages is not the same.

**Comment by PVince81** [comment]
can't use `$this->instanceOfStorage()` because `$this` is actually a `LocalRootStorage`...

I tried to also override `moveFromStorage` in `GroupFolderStorage` (which is the encrypted one) and make it use `copyFromStorage`, but there it also takes a shortcut: https://github.com/nextcloud/server/blob/stable24/lib/private/Files/Storage/Local.php#L563

and there I can't detect if `$this` because it's a `LocalRootStorage` (resolved from the `Jail` wrapper from the group folder)

this gets complicated :face_with_head_bandage:

**Comment by kleener** [comment]
Moin
so I have now version 25.0.2 and Groupfolders 13.1.0.
I have run the following command: 
occ config:app:set groupfolders enable_encryption --value=true

Unfortunately, files I move from my private folder to the group folder are still destroyed.
Do I need to clear a PHP cache first or did I forget something?

**Comment by PVince81** [comment]
@kleener no, this is an actual bug where it forgets to move the encryption keys

you can try moving the files back to the previous location outside of group folder so that it will find the keys again

**Comment by PVince81** [comment]
I heard from @come-nc / @CarlSchwan that there was also known issues where versions are also not moved properly, it is likely that the root cause is the same

**Comment by LivOriona** [comment]
I encountered the same problem as described, and here is a workaround, in order to move encrypted documents out of group folders into a group folder: 
* use and install the Nextcloud synchronization client (tested on Windows 11)
* copy or download the files (which are out of a group folder) from the PC file explorer to a group folder
* make sure that the client has done the synchronization
* reload the web page of Nextcloud and the files are then always editable

**Comment by venanj** [comment]
I got around this problem by keeping server-side encryption enabled, but disabling Encrypt home storage folder

![image](https://user-images.githubusercontent.com/42942113/220650278-9817cdbe-e366-4bc4-b945-6bbd48b113dd.png)

**Comment by TPrieser** [comment]
> I got around this problem by keeping server-side encryption enabled, but disabling Encrypt home storage folder
> 
> ![image](https://user-images.githubusercontent.com/42942113/220650278-9817cdbe-e366-4bc4-b945-6bbd48b113dd.png)

Seems like this also solved our problem

**Comment by kleener** [comment]
Since the last Nextcloud version 25.0.4 and Groupfolder 13.1.1 I could not reproduce the problem anymore

**Comment by PVince81** [comment]
@icewind1991 can you confirm whether all the related fixes were merged ? this ticket can then be closed

**Comment by madsonsouza22** [comment]
Hi, has anyone managed to solve the root problem? I'm on version 27 and I still have problems copying files to shared folders via the web interface
